<a href="https://colab.research.google.com/github/NarayaniSuresh/ADVANCEDDL/blob/main/BATCH_NORMALIZATION_E0321036.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# IMPORT LIBRARIES

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import cv2
import matplotlib.pyplot as plt
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, BatchNormalization, Activation, Dropout
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.callbacks import EarlyStopping
from tensorflow.keras import layers, models, datasets
from tensorflow.keras.utils import to_categorical

# LOAD DATA

In [ ]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()

11490434/11490434 [==============================] - 0s 0us/step


In [ ]:
X_train = X_train / 255.0
X_test = X_test / 255.0

In [ ]:
y_train = tf.keras.utils.to_categorical(y_train, 10)
y_test = tf.keras.utils.to_categorical(y_test, 10)

# BATCH NORMALISATION - AFTER THE ACTIVATION FUNCTION

In [ ]:
model2= models.Sequential([
    layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    layers.BatchNormalization(),
    layers.MaxPooling2D((2, 2)),
    layers.Conv2D(64, (3, 3), activation='relu'),
    layers.BatchNormalization(),
    layers.MaxPooling2D((2, 2)),
    layers.Conv2D(64, (3, 3), activation='relu'),
    layers.BatchNormalization(),
    layers.Flatten(),
    layers.Dense(64, activation='relu'),
    layers.BatchNormalization(),
    layers.Dense(10, activation='softmax')
])

In [ ]:
model2.compile(optimizer=Adam(),
              loss='categorical_crossentropy',
              metrics=['accuracy'])

In [ ]:
history2 = model2.fit(X_train, y_train, epochs=3, batch_size=128,
                      validation_data=(X_test, y_test), verbose=2)


Epoch 1/3
469/469 - 74s - loss: 0.1200 - accuracy: 0.9654 - val_loss: 0.4336 - val_accuracy: 0.8552 - 74s/epoch - 157ms/step
Epoch 2/3
469/469 - 67s - loss: 0.0360 - accuracy: 0.9897 - val_loss: 0.0385 - val_accuracy: 0.9880 - 67s/epoch - 143ms/step
Epoch 3/3
469/469 - 67s - loss: 0.0250 - accuracy: 0.9924 - val_loss: 0.0367 - val_accuracy: 0.9883 - 67s/epoch - 143ms/step


# BATCH SIZE - 32

In [ ]:
model.fit(X_train, y_train, epochs=5, batch_size=32, validation_data=(X_test, y_test))

Epoch 1/5
1875/1875 [==============================] - 87s 45ms/step - loss: 0.1046 - accuracy: 0.9687 - val_loss: 0.0424 - val_accuracy: 0.9867
Epoch 2/5
1875/1875 [==============================] - 85s 45ms/step - loss: 0.0450 - accuracy: 0.9860 - val_loss: 0.0485 - val_accuracy: 0.9838
Epoch 3/5
1875/1875 [==============================] - 80s 43ms/step - loss: 0.0321 - accuracy: 0.9896 - val_loss: 0.0346 - val_accuracy: 0.9887
Epoch 4/5
1875/1875 [==============================] - 81s 43ms/step - loss: 0.0256 - accuracy: 0.9921 - val_loss: 0.0264 - val_accuracy: 0.9911
Epoch 5/5
1875/1875 [==============================] - 84s 45ms/step - loss: 0.0221 - accuracy: 0.9930 - val_loss: 0.0385 - val_accuracy: 0.9884


In [ ]:
test_loss, test_acc = model.evaluate(X_test, y_test)
print("Test accuracy:", test_acc)

313/313 [==============================] - 4s 12ms/step - loss: 0.0385 - accuracy: 0.9884
Test accuracy: 0.9883999824523926


# BATCH SIZE - 64

In [ ]:
model.fit(X_train, y_train, epochs=5, batch_size=64, validation_data=(X_test, y_test))


Epoch 1/5
938/938 [==============================] - 80s 84ms/step - loss: 0.0084 - accuracy: 0.9974 - val_loss: 0.0235 - val_accuracy: 0.9924
Epoch 2/5
938/938 [==============================] - 73s 78ms/step - loss: 0.0046 - accuracy: 0.9987 - val_loss: 0.0224 - val_accuracy: 0.9937
Epoch 3/5
938/938 [==============================] - 76s 82ms/step - loss: 0.0064 - accuracy: 0.9980 - val_loss: 0.0304 - val_accuracy: 0.9910
Epoch 4/5
938/938 [==============================] - 77s 82ms/step - loss: 0.0069 - accuracy: 0.9979 - val_loss: 0.0270 - val_accuracy: 0.9927
Epoch 5/5
938/938 [==============================] - 77s 82ms/step - loss: 0.0048 - accuracy: 0.9985 - val_loss: 0.0300 - val_accuracy: 0.9916


In [ ]:
test_loss, test_acc = model.evaluate(X_test, y_test)
print("Test accuracy:", test_acc)

313/313 [==============================] - 4s 14ms/step - loss: 0.0300 - accuracy: 0.9916
Test accuracy: 0.991599977016449


# BATCH SIZE - 128

In [ ]:
model.fit(X_train, y_train, epochs=5, batch_size=128, validation_data=(X_test, y_test))

Epoch 1/5
469/469 [==============================] - 70s 150ms/step - loss: 0.0015 - accuracy: 0.9996 - val_loss: 0.0245 - val_accuracy: 0.9939
Epoch 2/5
469/469 [==============================] - 71s 152ms/step - loss: 6.3726e-04 - accuracy: 0.9998 - val_loss: 0.0238 - val_accuracy: 0.9940
Epoch 3/5
469/469 [==============================] - 72s 154ms/step - loss: 2.7737e-04 - accuracy: 1.0000 - val_loss: 0.0245 - val_accuracy: 0.9940
Epoch 4/5
469/469 [==============================] - 73s 156ms/step - loss: 2.0158e-04 - accuracy: 1.0000 - val_loss: 0.0230 - val_accuracy: 0.9942
Epoch 5/5
469/469 [==============================] - 70s 149ms/step - loss: 1.3983e-04 - accuracy: 1.0000 - val_loss: 0.0234 - val_accuracy: 0.9947


In [ ]:
test_loss, test_acc = model.evaluate(X_test, y_test)
print("Test accuracy:", test_acc)

313/313 [==============================] - 5s 14ms/step - loss: 0.0234 - accuracy: 0.9947
Test accuracy: 0.994700014591217


# BATCH SIZE - 256

In [ ]:
model.fit(X_train, y_train, epochs=5, batch_size=256, validation_data=(X_test, y_test))

Epoch 1/5
235/235 [==============================] - 72s 308ms/step - loss: 7.0096e-05 - accuracy: 1.0000 - val_loss: 0.0238 - val_accuracy: 0.9944
Epoch 2/5
235/235 [==============================] - 71s 304ms/step - loss: 6.3991e-05 - accuracy: 1.0000 - val_loss: 0.0242 - val_accuracy: 0.9944
Epoch 3/5
235/235 [==============================] - 73s 312ms/step - loss: 5.7032e-05 - accuracy: 1.0000 - val_loss: 0.0242 - val_accuracy: 0.9945
Epoch 4/5
235/235 [==============================] - 75s 318ms/step - loss: 5.2364e-05 - accuracy: 1.0000 - val_loss: 0.0241 - val_accuracy: 0.9944
Epoch 5/5
235/235 [==============================] - 78s 328ms/step - loss: 4.3433e-05 - accuracy: 1.0000 - val_loss: 0.0242 - val_accuracy: 0.9946


In [ ]:
test_loss, test_acc = model.evaluate(X_test, y_test)
print("Test accuracy:", test_acc)

313/313 [==============================] - 6s 18ms/step - loss: 0.0247 - accuracy: 0.9947
Test accuracy: 0.994700014591217


# BATCH SIZE - 512

In [ ]:
model.fit(X_train, y_train, epochs=5, batch_size=512, validation_data=(X_test, y_test))

Epoch 1/5
118/118 [==============================] - 89s 759ms/step - loss: 3.4615e-05 - accuracy: 1.0000 - val_loss: 0.0243 - val_accuracy: 0.9946
Epoch 2/5
118/118 [==============================] - 94s 794ms/step - loss: 3.2118e-05 - accuracy: 1.0000 - val_loss: 0.0245 - val_accuracy: 0.9947
Epoch 3/5
118/118 [==============================] - 83s 701ms/step - loss: 3.0334e-05 - accuracy: 1.0000 - val_loss: 0.0245 - val_accuracy: 0.9947
Epoch 4/5
118/118 [==============================] - 78s 657ms/step - loss: 2.8764e-05 - accuracy: 1.0000 - val_loss: 0.0247 - val_accuracy: 0.9947
Epoch 5/5
118/118 [==============================] - 76s 646ms/step - loss: 2.7454e-05 - accuracy: 1.0000 - val_loss: 0.0247 - val_accuracy: 0.9947


In [ ]:
test_loss, test_acc = model.evaluate(X_test, y_test)
print("Test accuracy:", test_acc)

313/313 [==============================] - 4s 12ms/step - loss: 0.0247 - accuracy: 0.9947
Test accuracy: 0.994700014591217


The evaluation of different batch sizes on the model's performance shows a clear trend in accuracy and loss metrics.

1. **Batch Size 32**: The model achieved an accuracy of 0.9884 with a loss of 0.0385. While this is a strong performance, it is the lowest among the tested batch sizes.

2. **Batch Size 64**: Increasing the batch size to 64 resulted in an improvement, with an accuracy of 0.9916 and a reduced loss of 0.0300. This indicates better model generalization and performance.

3. **Batch Size 128**: Further increasing the batch size to 128 led to a significant improvement, with an accuracy of 0.9947 and a loss of 0.0234. This suggests that the model benefits from a larger batch size, improving its predictive power.

4. **Batch Size 256**: The performance remains consistent with a batch size of 256, maintaining an accuracy of 0.9947 and a loss of 0.0247. This implies that the model's performance has plateaued at this batch size.

5. **Batch Size 512**: Similarly, with a batch size of 512, the model maintains its accuracy at 0.9947 with a loss of 0.0247, indicating that further increasing the batch size does not result in additional performance gains.

In conclusion, the model's performance improves with an increasing batch size up to 128, after which it plateaus. The optimal batch size for this model appears to be 128, providing the highest accuracy with the lowest loss. Further increases in batch size do not yield significant improvements, suggesting a point of diminishing returns.
For this specific model, using a batch size of 128 is recommended to achieve optimal performance. Further increases in batch size do not yield significant performance benefits, highlighting the importance of balancing batch size with computational efficiency.